### PR TITLE
Add DS-Chat comparison 

### PR DESCRIPTION
**Pull request by cat-state**
This PR optimizes the implementation of NeMo-PPO, adding hydra and bug fixes.

### CODE REVIEW

File: trlx/models/modeling_nemo_ppo.py [hunk 1]
```diff
@@ -191,93 +169,114 @@ def __init__(self, language_model, other_heads, build_reference_model=True):
...
-            sequence_parallel=self._lm.sequence_parallel,
-            gradient_accumulation_fusion=self._lm.gradient_accumulation_fusion,
-        )
+        logit_weights = self._lm.word_embeddings_weight()
 
-        if get_key_value:
-            logits, presents = logits
-            lm_output, lm_output_presents = lm_output
+        if run_policy_model:
+            self.offload_reference_model()
```
Comment (Dahoas) [inline trlx/models/modeling_nemo_ppo.py:238]: Does the reference model get offloaded more than once per execution?

Comment (Dahoas) [reply trlx/models/modeling_nemo_ppo.py:238]: Nevermind I see

File: trlx/models/modeling_nemo_ppo.py [hunk 2]
```diff
@@ -191,93 +169,114 @@ def __init__(self, language_model, other_heads, build_reference_model=True):
...
+                heads_output = None
         else:
+            logits = None
             heads_output = None
 
         if run_reference_model:
-            device = lm_output.device
-            self.reference_model.model.to(device, non_blocking=True)
-            ref_lm_output = self.reference_model.model(*args, get_key_value=get_key_value, **kwargs)
+            self.offload_policy_model()
```
Comment (Dahoas) [inline trlx/models/modeling_nemo_ppo.py:267]: How efficient is this? I'm curious if it gives a big slowdown

File: trlx/models/modeling_nemo_ppo.py [hunk 3]
```diff
@@ -494,7 +493,12 @@ def freeze_layers(m):
                         m.eval()
                         for p in m.parameters():
                             p.requires_grad_(False)
+                    else:
+                        for p in m.parameters():
+                            p.requires_grad_(True)
 
+            for p in gpt.parameters():
```
Comment (Dahoas) [inline trlx/models/modeling_nemo_ppo.py:500]: Is this to make sure the embedding layers get frozen?

**Review by Dahoas — COMMENTED** [review]
Looks good overall! Just left some comments on parts I looked closely at and didn't quite understand

File: trlx/models/modeling_nemo_ppo.py (hunk 1 above)
Comment (cat-state) [reply trlx/models/modeling_nemo_ppo.py:238]: Ah it should be memozied offloading

**Comment by Dahoas** [comment]
@cat-state I'm happy to merge this today if there are no other changes from your side

File: trlx/models/modeling_nemo_ppo.py [hunk 4]
```diff
@@ -191,93 +169,114 @@ def __init__(self, language_model, other_heads, build_reference_model=True):
...
-            self.reference_model = OffloadedModel(deepcopy(language_model), device=torch.cuda.current_device())
+            for p in self.language_model.parameters():
+                if p.requires_grad:
+                    p._ref_data = p.data.clone().cpu().pin_memory()
+                    p._policy_data = p.data
+
+            self.reference_model_offloaded = True
         else:
-            self.reference_model = None
+            self.reference_model_offloaded = True
```
[the remark this answers is not in the document]
Comment (cat-state) [reply trlx/models/modeling_nemo_ppo.py:181]: I think it doesn't matter either way

**Comment by cat-state** [comment]
> @cat-state I'm happy to merge this today if there are no other changes from your side

Yup, I don't have any changes on my side

**Review by Dahoas — APPROVED** [review]

**Pull request merged**